src/Table.js

```diff
@@ -0,0 +1,189 @@
+                    placeholder="Search..." 
+                    value={filterText}
+                    onChange={props.onChangeFilterText}
+                />
+            </div>
+            <div className="search-box">
+                <input 
+                    type="checkbox"
+                    id="in-stock-only"
+                    checked={inStockOnly}
+                    onChange={props.onChangeInStockOnly}
+                />
+                <label htmlFor="in-stock-only" >Only show products in stock</label>
+            </div>
+
+        </div>
+    );
+}
+
+class FilterableProductTable extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            inStockOnly: false,
+            filterText: "",
+        }
+        this.handleFilterText = this.handleFilterText.bind(this);
+        this.handleInStockOnly = this.handleInStockOnly.bind(this);
+    }
+
+    handleFilterText(e) {
+        this.setState({
+            filterText: e.target.value,
+        });
+    }
+
+    handleInStockOnly(e) {
+        this.setState({
+            inStockOnly: e.target.checked,
+        });
+    }
+
+    render() {
+        const inStockOnly = this.state.inStockOnly;
+        const filterText = this.state.filterText;
+        return(
+            <div className="main">
+                <SearchBar 
+                    inStockOnly={inStockOnly} 
+                    filterText={filterText}
+                    onChangeFilterText={this.handleFilterText}
+                    onChangeInStockOnly={this.handleInStockOnly}
+                />
+                <ProductTable 
+                    inStockOnly={inStockOnly} 
+                    filterText={filterText}
+                />
+            </div>
+        );
+    }
+}
+
+export default FilterableProductTable;
+
+/*
+function filterTable(table, filterText, inStockOnly) {
+    const path = new RegExp(filterText, "i");
+    let newTable = table.slice();
+    let stockTable = [];
+    let returnTable = [];
+    if(inStockOnly) {
+     newTable.map((value) => {
+         return (stockTable = value.stocked ? 
+        stockTable.concat(value) : 
+        stockTable);
+         });
+    } else {
+        stockTable = newTable;
+    }
+    stockTable.map((value, index) => {
+        return (returnTable = path.test(value.name) ? 
+            returnTable.concat(value) :
+            returnTable);
+    });
+    console.log("newTable: " + newTable);
+    console.log("stockTable: " + stockTable + ", inStock: " + inStockOnly);
+    console.log("returnTable: " + returnTable);
+    return returnTable.slice();  
+}*/
+
+function filterTable(table, filterText, inStockOnly) {
+    const path = new RegExp(filterText, "i");
+    let newTable = table.slice();
+    if(inStockOnly) {
+        newTable = newTable.filter(value => value.stocked); 
+    }
+    newTable = newTable.filter(value => path.test(value.name));
+    return newTable.slice();
```

Comment:
niepotrzebnie kopiujesz tablicę przed jej zwróceniem 
Poza tym Twój kod jest ok, ale napiszę pod  spodem jak ja bym to zrobił
`function filterTable(table, filterText, inStockOnly) {
    const path = new RegExp(filterText, "i");
    return table.filter(({name, stocked}) => inStockOnly ? (stocked && path.test(name)) : path.test(name));
}`